src/esi/Makefile.am

```diff
@@ -12,6 +12,11 @@ noinst_LTLIBRARIES = libesi.la
 
 ESI_PARSER_SOURCES =
 
+##
+## List in reverse order of preference.
+## The last parser to initialize will be used as default.
```

Comment:
We cannot guarantee/control static initialization order across translation units. AFAICT, in the current PR code, the static initialization order[^1] determines the order of the following two calls:

1. ESIParser::Register("expat", &ESIExpatParser::NewParser)
2. ESIParser::Register("libxml2", &ESILibxml2Parser::NewParser)

The fact that (some specific) initialization order is usually _preserved_ across builds on a given platform is irrelevant.

[^1]: The order of items in (presumably) the `ESI_PARSER_SOURCES` "list" in this Makefile is not necessarily the (static) initialization order.

Reply:
FTR; This comment is retained more as an advisory about implicit behaviour than a guarantee of which parser will be used.

Reply:
> FTR; This comment is retained

FTR: This comment is not "retained". The comment _text_ was _moved_ from the context where it could have been actually correct (at the very least, it made sense) to the context where its most natural interpretation would mislead the reader that the listed order controls which parser is going to be used by default.

> more as an advisory about implicit behaviour than a guarantee of which parser will be used.

If this is a comment about Squid behavior (i.e. Squid would use the last parser listed as the default parser when given a choice), then it is lying (i.e. when given a choice, Squid may use a different default parser instead). This Makefile list does not control implicit Squid behavior.

> I mean, this is _our_ preference of library. Vendors will link against their own choices.

If this comment is about Squid Project preference (whatever that means), then it should be rephrased to detail what that "preference" means (to the point where a reader can actually understand what we are talking about) and, more importantly, to remove the implication that the listed order actually controls which parser will be used as the default parser (when more then two are available).

Reply:
However, once the initialization order dependency bug is fixed, the comment should probably go back to the initialization code, where the old text will make sense again -- no pressing need to change the original comment text if its "original habitat" is effectively restored:

```C++
    // register in reverse order of preference.
    // The latest registered parser will be used as default.
```

Reply:
Please clarify what you mean by "the initialization order dependency bug".

The closest thing to a bug I see with ESI initiailization order, is that they do not report clearly which one will be used in absence of admin configuration. IIRC, this "default order" was an arbitrary maintainer decision based on which library had/has the most popularity and least security issues last time it was reviewed.

Reply:
> Please clarify what you mean by "the initialization order dependency bug".

Assuming both libraries are enabled/configured and `esi_parser` is set to `auto`, The ESIParser::NewParser() result in this PR is determined by static initialization order across two different translation units (i.e. by compiler, linker, and/or similar external and possibly random factors) rather than by Squid source code. More specifically,

* In the official code, the "auto" parser is ESIExpatParser[^1].
* In the PR code, it is impossible to reliably predict what the "auto" parser is based on the Squid source code tarball alone. That decision is no longer made by Squid code. It is made by static initialization order across translation units which is outside of our control. This is a bug.

[^1]: Yes, I know that cf.data.pre and convoluted Modules.cc documentation say it should be ESILibxml2Parser instead. If I am right, then this is a bug in the official code, introduced by 57b6788. If I am wrong, it will be ESILibxml2Parser instead. However, as far as _this_ PR is concerned, the bug is _not_ in the specific value returned by ESIParser::NewParser() when `esi_parser` is set to `auto`. I am not asking you to fix any old bugs here! The new bug introduced by this PR is that the return value became unpredictable, dependent on static initialization order across translation units, random, not controlled by Squid sources, whatever you want to call it.

Reply:
And see the [initial](https://github.com/squid-cache/squid/pull/965#discussion_r790299237) change request text for the calls which order was randomized by this PR.

Reply:
Hmm, I see. You are right there is a bug in the official code. I am have opened bug 5192 to fix that, when that is gone this change will be unnecessary.

Reply:
> when [Bug 5192] is gone this change will be unnecessary

That depends on how Bug 5192 is fixed. Minimal/surgical fixes of Big 5192 (e.g., replacing front() with back() in ESIParser::NewParser()) would not address _this_ change request because this change request is about a different bug introduced in this PR:

* Bug [5192](http://bugs.squid-cache.org/show_bug.cgi?id=5192): Documented order is the reverse of the actual order.
* This PR [bug](https://github.com/squid-cache/squid/pull/965#discussion_r790299237): The actual order is random/unpredictable/global-initialization-dependent/whatever-you-want-to-call-it.

Reply:
> Minimal/surgical fixes of Big 5192 would not address this change request

In fact, I see no practical way to leave this PR code "as is" after _any_ Bug 5192 fix, no matter how advanced. Some changes will be needed here. What kind of changes depends on how Bug 5192 is fixed.

Reply:
see PR #968.

Reply:
I reviewed PR 968. As I [expected](https://github.com/squid-cache/squid/pull/965#discussion_r791172205), that PR does not address this change request. After PR 968 lands, addressing this change request would require a complete removal of the PR-added comment this change request is attached to.